hw/ip/spi_host/doc/_index.md

```diff
@@ -0,0 +1,633 @@
+],
+  ["ByteOrder=1",
+  {name: "sd[0] (host output)", wave: "x22222222222|2222|222|22x", data: ["t[7]", "t[6]", "t[5]", "t[4]", "t[3]", "t[2]", "t[1]", "t[0]", "t[15]","t[14]",
+                                                                          "t[13]","t[9]","t[8]","t[23]","t[22]","t[16]", "t[31]", "t[30]", "t[25]", "t[24]"]},
+  {name: "sd[1] (host input)", wave: "x22222222222|2222|222|22x", data: ["r[7]", "r[6]", "r[5]", "r[4]", "r[3]", "r[2]", "r[1]", "r[0]", "r[15]","r[14]",
+                                                                         "r[13]","r[9]","r[8]","r[23]","r[22]","r[16]", "r[31]", "r[30]", "r[25]", "r[24]"]},
+  {name: "Which Byte?", wave: "x5.......5..|..5.|.5.|..x", data: ["LSB from TXDATA/RXDATA", "","", "          MSB"]}
+],
+  ],
+  head: {
+    text: "Serial bit ordering for 32-bit TXDATA & RXDATA data words (t[31:0] and r[31:0]), Standard SPI as a Function of the Parameter 'ByteOrder'",
+  },
+  foot: {
+  text: "(Bits are numbered as they appear when loaded into TXDATA or RXDATA registers.)"
+  }
+}
+{{< /wavejson >}}
+
+
+For Dual and Quad mode transfers similar time ordering applies, as shown in the following figure, though many bits of similar significance are packed into multiple parallel `sd` data lines, with the most significant going to `sd[0]` (*not* sd[3]).
+
+{{< wavejson >}}
+{signal: [
+  ["ByteOrder=0",
+  {name: "sd[0]", wave: "x...22334455x...", data: ["d[31]", "d[27]", "d[23]", "d[19]", "d[15]", "d[11]", "d[7]", "d[3]"]},
+  {name: "sd[1]", wave: "x...22334455x...", data: ["d[30]", "d[26]", "d[22]", "d[18]", "d[14]", "d[10]", "d[6]", "d[2]"]},
+  {name: "sd[2]", wave: "x...22334455x...", data: ["d[29]", "d[25]", "d[21]", "d[17]", "d[13]", "d[9]", "d[5]", "d[1]"]},
+  {name: "sd[3]", wave: "x...22334455x...", data: ["d[28]", "d[24]", "d[20]", "d[16]", "d[12]", "d[8]", "d[4]", "d[0]"]},
+],
+   ["ByteOrder=1",
+  {name: "sd[0]", wave: "x...55443322x...", data: ["d[7]", "d[3]", "d[15]", "d[11]", "d[23]", "d[19]", "d[31]", "d[27]"]},
+  {name: "sd[1]", wave: "x...55443322x...", data: ["d[6]", "d[2]", "d[14]", "d[10]", "d[22]", "d[18]", "d[30]", "d[26]"]},
+  {name: "sd[2]", wave: "x...55443322x...", data: ["d[5]", "d[1]", "d[13]", "d[9]", "d[21]", "d[17]", "d[29]", "d[25]"]},
+  {name: "sd[3]", wave: "x...55443322x...", data: ["d[4]", "d[0]", "d[12]", "d[8]", "d[20]", "d[16]", "d[28]", "d[24]"]},
+  ],
+  ],
+  head: {
+   text: "Serial bit ordering for 32-bit data word (d[31:0]), Quad SPI as a Function of the Parameter 'ByteOrder'",
+  },
+  foot: {
+  text: "(Bits are numbered as they appear when loaded into TXDATA or RXDATA registers.)"
+  }
+}
+{{< /wavejson >}}
+
+### Command Length and TXDATA/RXDATA Byte-masking (or Present Lack Thereof)
+
+There is no restriction on the length of a command.
+Even though the TXDATA and RXDATA registers require 4-byte accesses, any unused bytes in the TXDATA fifo (or unused slots in the RXDATA fifo), are simply ignored at the end of a command.
```

Comment:
hmmm.... this could end up becoming a problem. 
We can talk to the software guys a bit more..but i'm pretty sure they really don't like the idea of having to pack in software to 4-byte alignments. I'm pretty certain the ideal would be to treat this write almost like a memory.

Reply:
This section is going to be reworked in a PR that comes up very soon with "Programmer's Guide."
Some small amount of space wasted (max 3-bytes on either end) is going to be unavoidable if the bus width is 32-bits.

Making the fifo 8-bits is way simpler from a programmers perspective, one less later of byte ordering, but it's lame for time taken to load and un load the fifo.

I understand the concerns about bit-packing, however I would rather discuss the solution with software people directly.    I'm trying to provide the interface that I would prefer if I were writing the software, but there is room for variance of opinion.

I hope to have something written up soon for the programming model.

This will also fix the complete lack of byte-enables, which do in this exact commit requires some minimal packing. 
Adding byte-enables will get rid of that.

@gkelly, @a-pronin.

Reply:
In short though: Firmware developers do not need to pack bytes.
This is especially true in the nexte commit where TXDATA is a memory window

Reply:
yeah sounds good. Agreed that a minor waste of 3 bytes shouldn't be a big issue.
Btw, i'm not sure if this changes your default FIFO sizing a little, since I didn't think too closely about how the 264 is structured.

Reply:
Right now it is:
- 256 for data
- 1 word for instruction
- 1 word for address

**NOTE:** For 3-byte addressing mode (mandatory for smaller flash parts) instruction code should be loaded into address MSB. That is a packing restriction, but I can't see around it unless firmware puts address in one byte at a time.  In which case we have 12 wasted bytes.   This mode would bring the total count up 272.  If we do that, then that is probably as high as we need to go.

Reply:
I blew the TX FIFO size up to 256+32 = 288 bytes, because the largest possible TXOnly header is 5 bytes, and add to that 3 bytes for an unaligned TX, makes for 8 bytes of header, and then assuming worst cases scenario.

Shrinking the RX fifo to 256.

LMKWYT

Reply:
Also as far as the original alignment question is concerned this whole section is probably going to change as I write the Programmer's Guide.  For that reason I've moved this section to #5149.

This [comment is also referenced there](https://github.com/lowRISC/opentitan/pull/5149/files#r572616334)